src/sim/detail/InitializeTracks.cu

```diff
@@ -0,0 +1,361 @@
+struct IsEqual
+{
+    size_type value;
+
+    CELER_FUNCTION bool operator()(size_type x) const { return x == value; }
+};
+
+//---------------------------------------------------------------------------//
+// KERNELS
+//---------------------------------------------------------------------------//
+/*!
+ * Initialize the track states on device. The track initializers are created
+ * from either primary particles or secondaries. The new tracks are inserted
+ * into empty slots (vacancies) in the track vector.
+ */
+__global__ void init_tracks_kernel(const StatePointers            states,
+                                   const ParamPointers            params,
+                                   const TrackInitializerPointers inits)
+{
+    size_type count
+        = min<size_type>(inits.vacancies.size(), inits.initializers.size());
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < count)
+    {
+        // Get the track initializer from the back of the vector. Since new
+        // initializers are pushed to the back of the vector, these will be the
+        // most recently added and therefore the ones that still might have a
+        // parent they can copy the geometry state from.
+        const TrackInitializer& init
+            = inits.initializers[inits.initializers.size() - thread_id - 1];
+
+        // Index of the empty slot to create the new track in
+        ThreadId slot_id(
+            inits.vacancies[inits.vacancies.size() - thread_id - 1]);
+
+        // Initialize the simulation state
+        SimTrackView sim(states.sim, slot_id);
+        sim = init.sim;
+
+        // Initialize the particle physics data
+        ParticleTrackView particle(params.particle, states.particle, slot_id);
+        particle = init.particle;
+
+        // Copy the geometry state from the parent if possible
+        GeoTrackView geo(params.geo, states.geo, slot_id);
+        if (thread_id < inits.parent.size())
+        {
+            unsigned int parent_id
+                = inits.parent[inits.parent.size() - thread_id - 1];
+            GeoTrackView parent(params.geo, states.geo, ThreadId(parent_id));
+            geo = {parent, init.geo.dir};
+        }
+        // Initialize it from the position otherwise
+        else
+        {
+            geo = init.geo;
+        }
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find empty slots in the track vector and count the number of secondaries
+ * that survived cutoffs for each interaction. If the track is dead and
+ * produced secondaries, fill the empty track slot with one of the secondaries.
+ */
+__global__ void locate_alive_kernel(const StatePointers            states,
+                                    const ParamPointers            params,
+                                    const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
```

Comment:
In a case of "six of one, half a dozen of the other", it might be better to leave the `ThreadId` in its original type here:
```suggestion
    const auto thread_id = KernelParamCalculator::thread_id();
```
The `if` will still work, and below you'll just replace `thread_id` with `thread_id.get()`; and `ThreadId(thread_id)` with `thread_id`.

I think it's a bit "safer" this way because the type safety is only removed locally to where it's needed; and it's not much more verbose because of the number of places where the typed thread ID is actually used.

Reply:
Fair enough!